vowpalwabbit/slim/include/vw_slim_predict.h

```diff
@@ -0,0 +1,434 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <string>
+
+// avoid mmap dependency
+#define DISABLE_SHARED_WEIGHTS
+
+#include "example_predict.h"
+#include "explore.h"
+#include "gd_predict.h"
+#include "model_parser.h"
+#include "opts.h"
+
+namespace vw_slim {
+
+	/**
+	* @brief Exploration algorithm specified by the model.
+	*/
+	enum vw_predict_exploration
+	{
+		epsilon_greedy,
+		softmax,
+		bag
+	};
+
+	uint64_t ceil_log_2(uint64_t v);
+
+	// this guard assumes that namespaces are added in order
+	// the complete feature_space of the added namespace is cleared afterwards
+	class namespace_copy_guard
+	{
+		example_predict& _ex;
+		unsigned char _ns;
+		bool _remove_ns;
+	public:
+		namespace_copy_guard(example_predict& ex, unsigned char ns);
+		~namespace_copy_guard();
+
+		void feature_push_back(feature_value v, feature_index idx);
+	};
+
+
+	class feature_offset_guard
+	{
+		example_predict& _ex;
+		uint64_t _old_ft_offset;
+	public:
+		feature_offset_guard(example_predict& ex, uint64_t ft_offset);
+		~feature_offset_guard();
+	};
+
+	class stride_shift_guard
+	{
+		example_predict& _ex;
+		uint64_t _shift;
+	public:
+		stride_shift_guard(example_predict& ex, uint64_t shift);
+		~stride_shift_guard();
+	};
+
+	/**
+	* @brief Vowpal Wabbit slim predictor. Supports: regression, multi-class classification and contextual bandits.
+	*/
+	template<typename W>
+	class vw_predict
+	{
+		std::unique_ptr<W> _weights;
+		std::string _id;
+		std::string _version;
+		std::string _command_line_arguments;
+		std::vector<std::string> _interactions;
+		bool _ignore_linear[256];
```

Comment:
This should be a std::array<bool, 256>

Reply:
Perf or memset() vs std::fill() ?  

---
In reply to: [304043864](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304043864) [](ancestors = 304043864)

Reply:
The generated assembly is identical. Please see this godbolt: https://godbolt.org/z/pX2yZg

Reply:
Nice.  Sold.

---
In reply to: [304472237](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304472237) [](ancestors = 304472237)

Reply:
spoke too soon.  bool _ignore_linear[256]; needed for gd::inline_predict.  gd_predict.h

    score = GD::inline_predict<W>(*_weights, false, _ignore_linear, _interactions, /* permutations */ false, ex);



---
In reply to: [304473387](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304473387) [](ancestors = 304473387,304472237)

Reply:
As an aside msvc generates different code... 

---
In reply to: [304479885](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304479885) [](ancestors = 304479885,304473387,304472237)